File: src/WarungRafi.Core/Finance.cs
using System.Security.Cryptography;
using System.Text.Json;

namespace WarungRafi.Core;

public sealed record CashSession(string Id,DateTimeOffset OpenedAt,long OpeningCash,string OpenedBy,
    DateTimeOffset? ClosedAt=null,long? CountedCash=null,long? ExpectedAtClose=null,string ClosingNote="")
{
    public long? Difference=>CountedCash-ExpectedAtClose;
}
public sealed record CashPosition(CashSession Session,long CashSales,long QrisSales,long CashIn,long CashOut,long CashRefunds,long SalesRefunds,long SaleCount)
{
    public long Expected=>checked(Session.OpeningCash+CashSales+CashIn-CashOut-CashRefunds);
    public long Gross=>checked(CashSales+QrisSales);
}
public enum RefundState { Requested,Completed,Failed }
public enum RefundChannel { Cash,External }
public sealed record Refund(string Id,string OrderId,long Amount,RefundChannel Channel,string Reason,
    RefundState State,DateTimeOffset RequestedAt,DateTimeOffset? CompletedAt=null,string? SessionId=null,
    string ApprovedBy="",string Reference="",string FailureReason="");
public sealed record FinanceEvent(string Id,long Sequence,string Kind,DateTimeOffset OccurredAt,
    string? SessionId,string? OrderId,long Amount,long CashDelta,string Actor,string Reason,JsonElement Details);

public static class FinanceRules
{
    public static void Amount(long value,bool allowZero=false)
    { if(value<(allowZero?0:1)||value>1_000_000_000)throw new ArgumentException("Nominal harus berupa rupiah bulat antara "+(allowZero?"0":"1")+" dan 1 miliar."); }
    public static void Id(string value)
    { if(!Guid.TryParseExact(value,"N",out _))throw new ArgumentException("Identitas tindakan tidak valid."); }
    public static string Note(string value)
    { var clean=value.Trim();if(clean.Length is <3 or >200)throw new ArgumentException("Isi keterangan 3–200 karakter.");return clean; }
    public static long EstimateFee(long amount,int basisPoints)
    {
        if(amount<=0||amount>999_999_999_999||basisPoints is <0 or >10000)throw new ArgumentException("Tarif/nominal tidak valid.");
        return checked((long)decimal.Round(amount*(decimal)basisPoints/10000,0,MidpointRounding.AwayFromZero));
    }
}

public static class ManagerPin
{
    public static string Hash(string pin)
    {
        if(pin.Length is <6 or >12 || pin.Any(c=>!char.IsAsciiDigit(c)))throw new ArgumentException("PIN harus 6–12 angka.");
        var salt=RandomNumberGenerator.GetBytes(16);
        var hash=Rfc2898DeriveBytes.Pbkdf2(pin,salt,600000,HashAlgorithmName.SHA256,32);
        return $"pbkdf2-sha256$600000${Convert.ToBase64String(salt)}${Convert.ToBase64String(hash)}";
    }
    public static bool Verify(string? encoded,string pin)
    {
        if(encoded is null||pin.Length is <6 or >12||pin.Any(c=>!char.IsAsciiDigit(c)))return false;
        try
        {
            var parts=encoded.Split('$');
            if(parts.Length!=4||parts[0]!="pbkdf2-sha256"||parts[1]!="600000")return false;
            var salt=Convert.FromBase64String(parts[2]);var expected=Convert.FromBase64String(parts[3]);
            if(salt.Length!=16||expected.Length!=32)return false;
            return CryptographicOperations.FixedTimeEquals(expected,Rfc2898DeriveBytes.Pbkdf2(pin,salt,600000,HashAlgorithmName.SHA256,32));
        }
        catch(FormatException){return false;}
    }
}

File: apps/admin/lib/finance.ts
import { HttpError, object } from "./http.ts";

const invalid = () => new HttpError(400, "Catatan keuangan tidak valid.");
const integer = (value: unknown, min = 0, max = Number.MAX_SAFE_INTEGER) => Number.isSafeInteger(value) && Number(value) >= min && Number(value) <= max;
const guid = (value: unknown) => typeof value === "string" && /^[a-f0-9]{32}$/.test(value);
const date = (value: unknown) => typeof value === "string" && value.length <= 40 && Number.isFinite(Date.parse(value));
const note = (value: unknown, min = 0) => typeof value === "string" && value.trim().length >= min && value.length <= 200;
const kinds = new Set(["session_opened", "session_closed", "sale", "cash_in", "cash_out", "refund_requested", "refund_completed", "refund_failed"]);
export type FinanceEvent = { id: string; sequence: number; kind: string; occurredAt: string; sessionId: string | null; orderId: string | null; amount: number; cashDelta: number; actor: string; reason: string; details: Record<string, unknown> };

// Shape validation at the HTTP boundary; PostgreSQL validates ordered state transitions.
export function financeEvents(input: unknown): FinanceEvent[] {
  if (!Array.isArray(input) || input.length < 1 || input.length > 50) throw invalid();
  return input.map(value => {
    const e = object(value); const d = object(e.details);
    if (typeof e.kind !== "string" || !kinds.has(e.kind) || !integer(e.sequence, 1) || !date(e.occurredAt) || !integer(e.amount) || !integer(e.cashDelta, -Number.MAX_SAFE_INTEGER) || !note(e.reason) || !["Kasir", "Pengelola"].includes(String(e.actor))) throw invalid();
    if (e.sessionId !== null && !guid(e.sessionId) || e.orderId !== null && !guid(e.orderId)) throw invalid();
    let id: string; let delta: number; const amount = Number(e.amount);
    if (e.kind === "session_opened" || e.kind === "session_closed") {
      if (!guid(e.sessionId) || e.orderId !== null || d.id !== e.sessionId || !date(d.openedAt) || d.openedBy !== "Kasir" || !integer(d.openingCash, 0, 1_000_000_000) || !note(d.closingNote)) throw invalid();
      const open = e.kind === "session_opened"; id = `${open ? "open" : "close"}-${e.sessionId}`; delta = open ? amount : 0;
      if (open ? amount !== d.openingCash || d.closedAt !== null || d.countedCash !== null || d.expectedAtClose !== null : !date(d.closedAt) || !integer(d.countedCash, 0, 1_000_000_000) || !integer(d.expectedAtClose) || amount !== d.countedCash || e.reason !== d.closingNote) throw invalid();
    } else if (e.kind === "sale") {
      if (!guid(e.sessionId) || !guid(e.orderId) || amount < 1 || d.orderId !== e.orderId || !integer(d.method, 0, 1) || d.amount !== amount || !integer(d.tendered, amount) || d.change !== Number(d.tendered) - amount || !date(d.paidAt) || d.method === 1 && (d.tendered !== amount || d.change !== 0)) throw invalid();
      id = `sale-${e.orderId}`; delta = d.method === 0 ? amount : 0;
    } else if (e.kind === "cash_in" || e.kind === "cash_out") {
      if (!guid(e.id) || !guid(e.sessionId) || e.orderId !== null || !integer(amount, 1, 1_000_000_000) || !note(e.reason, 3) || Object.keys(d).length !== 0) throw invalid();
      id = e.id as string; delta = e.kind === "cash_in" ? amount : -amount;
    } else {
      if (!guid(d.id) || !guid(e.orderId) || d.orderId !== e.orderId || d.amount !== amount || !integer(amount, 1, 1_000_000_000) || !integer(d.channel, 0, 1) || !note(d.reason, 3) || !date(d.requestedAt)) throw invalid();
      const requested = e.kind === "refund_requested"; const completed = e.kind === "refund_completed";
      id = `${requested ? "request" : "resolve"}-${d.id}`; delta = completed && d.channel === 0 ? -amount : 0;
      if (d.state !== (requested ? 0 : completed ? 1 : 2) || d.sessionId !== e.sessionId || (completed ? !guid(e.sessionId) || !date(d.completedAt) || !note(d.reference, 3) || d.failureReason !== "" : e.sessionId !== null || d.completedAt !== null)) throw invalid();
      if (requested ? e.actor !== "Kasir" || d.approvedBy !== "" || d.reference !== "" || d.failureReason !== "" || e.reason !== d.reason : e.actor !== "Pengelola" || d.approvedBy !== "Pengelola" || (completed ? e.reason !== d.reference : !note(d.failureReason, 3) || e.reason !== d.failureReason || d.reference !== "")) throw invalid();
    }
    if (e.id !== id || e.cashDelta !== delta) throw invalid();
    if (!e.kind.startsWith("refund_") && e.actor !== "Kasir") throw invalid();
    return { id, sequence: e.sequence as number, kind: e.kind, occurredAt: e.occurredAt as string, sessionId: e.sessionId as string | null, orderId: e.orderId as string | null, amount, cashDelta: delta, actor: e.actor as string, reason: e.reason as string, details: d };
  });
}
